src/jobwatcher/plugins/utils.py

```diff
@@ -8,50 +8,90 @@
 # or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
 # OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
 # limitations under the License.
+import copy
 import logging
 
 log = logging.getLogger(__name__)
 
 
-def get_optimal_nodes(nodes_requested, slots_requested, instance_properties):
+def get_optimal_nodes(nodes_requested, resources_requested, instance_properties):
     """
     Get the optimal number of nodes required to satisfy the number of nodes and slots requested.
 
     :param nodes_requested: Array containing the number of nodes requested by the ith job
-    :param slots_requested: Array containing the number of slots requested by the ith job
-    :param instance_properties: instance properties, i.e. number of slots available per node
+    :param resources_requested: Array containing dict of all resources requested by the ith job,
+        i.e. {"slots": 4, "gpus": 4}
+    :param instance_properties: instance properties, i.e. slots/gpu/memory available per node
     :return: The optimal number of nodes required to satisfy the input queue.
     """
-    vcpus = instance_properties.get("slots")
-    slots_remaining_per_node = []
-
-    for slots, num_of_nodes in zip(slots_requested, nodes_requested):
-        log.info("Requested %s nodes and %s slots" % (num_of_nodes, slots))
-        # For simplicity, uniformly distribute the numbers of cpus requested across all the requested nodes
-        slots_required_per_node = -(-slots // num_of_nodes)
-
-        if slots_required_per_node > vcpus:
-            log.warning(
-                "Slots required per node (%d) is greater than vcpus available on single node (%d), skipping job...",
-                slots_required_per_node,
-                vcpus,
+    resources_remaining_per_node = []
+
+    for job_resources, num_of_nodes in zip(resources_requested, nodes_requested):
+        log.info(
+            "Processing job that requested {0} nodes and the following resources: {1}".format(
+                num_of_nodes, job_resources
             )
+        )
+
+        # For simplicity, uniformly distribute the resource requested across all the requested nodes
+        job_resources_per_node = {}
+        for resource_type in job_resources:
+            job_resources_per_node[resource_type] = -(-job_resources[resource_type] // num_of_nodes)
+
+        job_valid = _job_runnable_on_given_node(job_resources_per_node, instance_properties, existing_node=False)
```

Comment:
Can we move this logic to get_pending_jobs_info together with the other checks on jobs requirements?